ramips: add support for Edimax CAP1200

eljefe80 opened the pull request:
Add the V7 APAC1200 build target

Signed-off-by: Jeff Williams <jeff@wdwconsulting.net>

obsy commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -36,6 +36,14 @@ define Device/alfa-network_ac1200rm
+define Device/apac1200
+  DTS := APAC1200
+  DEVICE_TITLE := V7 APAC1200
+  DEVICE_PACKAGES := kmod-ath10k kmod-ath10k-firmware-qca988x
```
kmod-ath10k-firmware-qca988x <- wrong package name. Should be ath10k-firmware-qca988x

mkresin commented:
Beside other issues, your commits have formal ones. Please follow our [submitting patches guidelines](https://lede-project.org/submitting-patches) and provide a hardware description and an install guide:

>If you add support for new hardware: Include in your commit message a short description of the >hardware and how to install LEDE on it. Have a look at the [recent additions](https://git.openwrt.org/?p=openwrt%2Fopenwrt.git&a=search&h=HEAD&st=commit&s=add+support+for) for some examples.

Without a hardware description it isn't possible to do a review. Please squash your commits into one as well.

pepe2k commented on target/linux/ramips/dts/APAC1200.dts:
```diff
@@ -0,0 +1,174 @@
+/*
+ *  BSD LICENSE
+ *
+ *  Copyright (C) 2018 Piotr Dymacz <pepe2k@gmail.com>
```
Please, use your copyright line here.

eljefe80 committed:
ramips: add support for Edimax CAP1200

The patch adds support for the Edimax CAP1200 and V7 APAC1200:

http://www.edimax.com/edimax/merchandise/merchandise_detail/data/edimax/global/edimax_pro_indoor_access_points_ac1200/cap1200/

and

http://www.v7world.com/uk/products/network/access-point/ac1200-dual-band-access-point.html

Specifications:
  * SoC: 580Mhz MT7620a
  * RAM: 64MB
  * Storage: 8MB of FLASH (SPI NOR)
  * 1x ar8035 GB switch
  * Ethernet: 1x1000M (802.11at POE), port #4 on ar8035
  * Wireless: QCA988X 802.11ac
              MT7620 2.4GHz 802.11b/g/n
  * 2x GPIO LED
  * 1x GPIO Reset Button
  * 1x DC Jack 12v
  * 1x UART, 3.3v, 57600

Installation:

   There are 2 ways of installation:

Serial Port/TFTP

1. Setup a tftp server and copy/link an initramfs image
 $ sudo ifconfig eth0 192.168.99.9 netmask 255.255.255.0
 $ cp openwrt-ramips-mt7620-edimax_cap1200-8m-initramfs-kernel.bin \
   /srv/tftp/uImage_EW-7476HPC

2. Connect via ttyl uart:
 On the board is a 4-pin, 3.3v uart running at 115200bps, 8n1, etc.

3. For a few seconds after power on and during reboot, you can hit 1 to
 break the automatic boot sequence and it will prompt for tftp settings,
 use the default. This will boot the openwrt initramfs kernel.

4. Change your ip address to 192.168.1.2
 $ sudo ifconfig eth0 192.168.1.2 netmask 255.255.255.0

5. Upload the sysupgrade image to the board with scp:
 $ scp openwrt-ramips-mt7620-edimax_cap1200-8m-squashfs-sysupgrade.bin \
 root@192.168.1.1:/tmp/fw.bin

6. Log in to the running system listening on 192.168.1.1 via ssh
 as root (without password):
 $ ssh root@192.168.1.1

7. Flash the uploaded firmware file from the ssh session via the
 sysupgrade command:
 root@OpenWrt:~# sysupgrade /tmp/fw.bin

Web interface

1. Navigate to the upgrade page on the device and upload the file
  openwrt-ramips-mt7620-edimax_cap1200-8m-squashfs-factory.bin

Signed-off-by: Jeff Williams <jeff@wdwconsulting.net>

mkresin commented on target/linux/ramips/base-files/etc/board.d/01_leds:
```diff
@@ -68,6 +68,10 @@ all0256n-8M)
 	ucidef_set_led_rssi "rssihigh" "RSSIHIGH" "all0256n:green:rssihigh" "wlan0" "70" "100" "-69" "8"
 	set_wifi_led "rt2800pci-phy0::radio"
 	;;
+apac1200)
```
use `v7,apac1200` as boardname. Please take care of tabs vs. spaces.

mkresin commented on target/linux/ramips/base-files/etc/board.d/02_network:
```diff
@@ -235,6 +235,7 @@ ramips_setup_interfaces()
 		ucidef_add_switch "switch0" \
 			"0:lan" "1:lan" "2:lan" "3:lan" "4:wan" "9@eth0"
 		;;
+        apac1200|\
```
use `v7,apac1200` as boardname. Please take care of tabs vs. spaces.

mkresin commented on target/linux/ramips/base-files/etc/board.d/02_network:
```diff
@@ -385,6 +386,9 @@ ramips_setup_macs()
 	wnce2001)
 		wan_mac=$(mtd_get_mac_binary factory 46)
 		;;
+        apac1200)
```
You can drop it here. You already have `mtd-mac-address = <&config 0x29>` in your dts, which should set the correct mac address. 

I'm not yet sure about the mac address increment. Here you don't have an increment, but in the dts you increment the base mac address by +3.

mkresin commented on target/linux/ramips/base-files/etc/hotplug.d/firmware/10-rt2x00-eeprom:
```diff
@@ -46,6 +46,11 @@ case "$FIRMWARE" in
 		rt2x00_eeprom_extract "factory" 0 512
 		rt2x00_eeprom_set_macaddr $wifi_mac
 		;;
+	apac1200)
```
use `v7,apac1200` as boardname. Please take care of tabs vs. spaces.

mkresin commented on target/linux/ramips/base-files/lib/ramips.sh:
```diff
@@ -49,6 +49,9 @@ ramips_board_detect() {
 	*"ALL5003")
 		name="all5003"
 		;;
+        *"APAC1200")
```
Drop it to use the generic board detection. The userspace board name will be `v7,apac1200`.

mkresin commented on target/linux/ramips/dts/APAC1200.dts:
```diff
@@ -0,0 +1,174 @@
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "V7,apac1200", "ralink,mt7620a-soc";
```
`v7,apac1200` please.

mkresin commented on target/linux/ramips/dts/APAC1200.dts:
```diff
@@ -0,0 +1,174 @@
+        };
+};
+
+
```
drop this blank line please. one is enough.

mkresin commented on target/linux/ramips/dts/APAC1200.dts:
```diff
@@ -0,0 +1,174 @@
+};
+
+&pinctrl {
+        state_default: pinctrl0 {
```
please use tabs instead of spaces.

mkresin commented on target/linux/ramips/dts/APAC1200.dts:
```diff
@@ -0,0 +1,174 @@
+        status = "okay";
+        pinctrl-names = "default";
+        pinctrl-0 = <&rgmii1_pins &rgmii2_pins &mdio_pins>;
+        mediatek,portmap = "wllll";
```
drop it. adding a default vlan map for a single port board doesn't make much sense.

mkresin commented on target/linux/ramips/dts/APAC1200.dts:
```diff
@@ -0,0 +1,174 @@
+        pinctrl-0 = <&rgmii1_pins &rgmii2_pins &mdio_pins>;
+        mediatek,portmap = "wllll";
+
+        port@4 {
```
are you sure it is required? what is attached to port 4? Is there an extra phy attached? Couldn't find anything related in your commit message.

The same applies to the mdio bus.

mkresin commented on target/linux/ramips/dts/APAC1200.dts:
```diff
@@ -0,0 +1,174 @@
+	status = "okay";
+
+	pcie-bridge {
+		mt76@0,0 {
```
I'm confused. According to your commit message you have a QCA988X (presumably attached via PCI) and the in-SoC mt7620 wireless.

But this whole node is about a MT76x2e, MT7603, MT7628 and MT7688 and you don't even include the corresponding mt76 driver in your image. Copy/Paste?

mkresin commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -36,6 +36,14 @@ define Device/alfa-network_ac1200rm
+  DTS := APAC1200
+  DEVICE_TITLE := V7 APAC1200
+  DEVICE_PACKAGES := kmod-ath10k ath10k-firmware-qca988x
+#  IMAGE/factory.bin := v7_encode.py -
```
drop it or build a factory image

mkresin replied:
please add `IMAGE_SIZE`, which should match the size of the firmware partition.

mkresin commented on target/linux/ramips/dts/APAC1200.dts:
```diff
@@ -0,0 +1,174 @@
+        };
+};
+
+&gsw {
```
Same question as above. Is some external phy/switch attached to port4 of the internal switch?

mkresin commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -36,6 +36,14 @@ define Device/alfa-network_ac1200rm
 endef
 TARGET_DEVICES += alfa-network_ac1200rm
 
+define Device/apac1200
```
please use ` Device/v7_apac1200` here.

mkresin commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -36,6 +36,14 @@ define Device/alfa-network_ac1200rm
+  DEVICE_PACKAGES := kmod-ath10k ath10k-firmware-qca988x
+#  IMAGE/factory.bin := v7_encode.py -
+endef
+TARGET_DEVICES += apac1200
```
`TARGET_DEVICES += v7_apac1200` please

mkresin commented on target/linux/ramips/dts/APAC1200.dts:
```diff
@@ -0,0 +1,174 @@
+                        reg = <4>;
+                        phy-mode = "rgmii";
+                };
+
```
drop this blank line please

mkresin commented on target/linux/ramips/base-files/etc/hotplug.d/firmware/10-rt2x00-eeprom:
```diff
@@ -46,6 +46,11 @@ case "$FIRMWARE" in
 		rt2x00_eeprom_set_macaddr $wifi_mac
 		;;
+	apac1200)
+		wifi_mac=$(mtd_get_mac_ascii config INIC_MAC_ADDR)
```
Something seem to me wrong. You are reading the mac address from the very same partition as binary in case of the ethernet mac address. But here you need to do an ascii read. Does the partition have ascii content or binary content?

Does the factory partition includes the mac address by any chance? would you please attach a dump of the factory partition.

mkresin commented on target/linux/ramips/dts/APAC1200.dts:
```diff
@@ -0,0 +1,174 @@
+
+&wmac {
+	status = "okay";
+        #mtd-mac-address = <&config 0x29>;
```
depending on some earlier questions, this need to be enabled or dropped.

mkresin commented:
> The calibration data for the QCA988X card is only available in a secondary squashfs filesystem, located at 0x71A000. This will be overwritten by either of the following processes
You can copy the file to /lib/firmware/ath10k/ after installation from https://github.com/eljefe80/openwrt-etc/raw/master/ramips-apac1200/cal-pci-0000:01:00.0.bin

Don't do it that way. The calibration data are individual per device. Add the partition holding the calibration data to the dts. It is most likely named "caldata" or "art" by the original firmware. Shrink the firmware partition accordingly.

Add a hotplug script to extract the firmware. The driver will send a request if the firmware is required, which will trigger the hotplug script. Have a look at https://github.com/lede-project/source/tree/master/target/linux/lantiq/base-files/etc/hotplug.d/firmware/ for existing files.

Please provide an bootlog from the original firmware. I would like ensure that there isn't more overwritten.

> Encode the sysupgrade image using the encode_v7.py tool here: https://raw.githubusercontent.com/eljefe80/openwrt-etc/master/ramips-apac1200/encode_v7.py

Please check if we already support this image format by any of the tools in https://github.com/openwrt/openwrt/tree/master/tools/firmware-utils/src. If not, add your own there. Check the existing files regarding how it should be done.

Please use `ramips: add support for V7 APAC1200` as commit title. Please break your lines in the commit message after 75 chars as mentioned in the submitting patches guideline.

mkresin reviewed (changes requested):
Please be prepared for more change requests. Given the number of issues I already spotted, I expect another round for this PR.

Please do a full test cycle if you have the requested changes done and upload a bootlog of an installed OpenWrt.

eljefe80 commented:
From the original firmware:

# dmesg                                                                         
Ralink MT7620   ver:2 eco:6 running at 580.00 MHz                               
console [ttyS1] enabled 

# cat /proc/mtd                                                                 
dev:    size   erasesize  name                                                  
mtd0: 00800000 00001000 "ALL"                                                   
mtd1: 00030000 00001000 "Bootloader"                                
mtd2: 00010000 00001000 "Config"                               
mtd3: 00008000 00001000 "Factory"                                    
mtd4: 00008000 00001000 "ART"                                    
mtd5: 006ca000 00001000 "Kernel"                                    
mtd6: 00564000 00001000 "user"
mtd7: 00096000 00001000 "manufacture"                                     
mtd8: 00010000 00001000 "backup"                                     
mtd9: 00040000 00001000 "storage"

eljefe80 commented:
From an install, using this pull request:

`BusyBox v1.27.2 () built-in shell (ash)

  _______                     ________        __
 |       |.-----.-----.-----.|  |  |  |.----.|  |_
 |   -   ||  _  |  -__|     ||  |  |  ||   _||   _|
 |_______||   __|_____|__|__||________||__|  |____|
          |__| W I R E L E S S   F R E E D O M
 -----------------------------------------------------
 OpenWrt SNAPSHOT, r5973-2dd0f32167
 -----------------------------------------------------
=== WARNING! =====================================
There is no root password defined on this device!
Use the "passwd" command to set up a new password
in order to prevent unauthorized SSH logins.
--------------------------------------------------
root@OpenWrt:~# dmesg
[    0.000000] Linux version 4.9.77 (jeff@maria) (gcc version 5.5.0 (OpenWrt GCC 5.5.0 r5949-dc555d003c) ) #0 Tue Jan 30 00:22:07 2018
[    0.000000] Board has DDR2
[    0.000000] Analog PMU set to hw control
[    0.000000] Digital PMU set to hw control
[    0.000000] SoC Type: MediaTek MT7620A ver:2 eco:6
[    0.000000] bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 00019650 (MIPS 24KEc)
[    0.000000] MIPS: machine is V7 APAC1200
[    0.000000] Determined physical RAM map:
[    0.000000]  memory: 04000000 @ 00000000 (usable)
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] On node 0 totalpages: 16384
[    0.000000] free_area_init_node: node 0, pgdat 803f2e94, node_mem_map 81000020
[    0.000000]   Normal zone: 128 pages used for memmap
[    0.000000]   Normal zone: 0 pages reserved
[    0.000000]   Normal zone: 16384 pages, LIFO batch:3
[    0.000000] pcpu-alloc: s0 r0 d32768 u32768 alloc=1*32768
[    0.000000] pcpu-alloc: [0] 0 
[    0.000000] Built 1 zonelists in Zone order, mobility grouping on.  Total pages: 16256
[    0.000000] Kernel command line: console=ttyS0,57600 rootfstype=squashfs,jffs2
[    0.000000] PID hash table entries: 256 (order: -2, 1024 bytes)
[    0.000000] Dentry cache hash table entries: 8192 (order: 3, 32768 bytes)
[    0.000000] Inode-cache hash table entries: 4096 (order: 2, 16384 bytes)
[    0.000000] Writing ErrCtl register=0001c40c
[    0.000000] Readback ErrCtl register=0001c40c
[    0.000000] Memory: 60428K/65536K available (3144K kernel code, 152K rwdata, 752K rodata, 168K init, 216K bss, 5108K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS:256
[    0.000000] CPU Clock: 580MHz
[    0.000000] clocksource: systick: mask: 0xffff max_cycles: 0xffff, max_idle_ns: 583261500 ns
[    0.000000] systick: enable autosleep mode
[    0.000000] systick: running - mult: 214748, shift: 32
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 6590553264 ns
[    0.000011] sched_clock: 32 bits at 290MHz, resolution 3ns, wraps every 7405115902ns
[    0.015457] Calibrating delay loop... 385.84 BogoMIPS (lpj=1929216)
[    0.080610] pid_max: default: 32768 minimum: 301
[    0.089915] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.102947] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.123760] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.143244] futex hash table entries: 256 (order: -1, 3072 bytes)
[    0.155386] pinctrl core: initialized pinctrl subsystem
[    0.166354] NET: Registered protocol family 16
[    0.658111] PCI host bridge /pcie@10140000 ranges:
[    0.667507]  MEM 0x0000000020000000..0x000000002fffffff
[    0.677880]   IO 0x0000000010160000..0x000000001016ffff
[    0.700987] rt2880_gpio 10000600.gpio: registering 24 gpios
[    0.712018] rt2880_gpio 10000600.gpio: registering 24 irq handlers
[    0.725142] PCI host bridge to bus 0000:00
[    0.733140] pci_bus 0000:00: root bus resource [mem 0x20000000-0x2fffffff]
[    0.746855] pci_bus 0000:00: root bus resource [io  0xffffffff]
[    0.758584] pci_bus 0000:00: root bus resource [??? 0x00000000 flags 0x0]
[    0.772086] pci_bus 0000:00: No busn resource found for root bus, will use [bus 00-ff]
[    0.787874] pci 0000:00:00.0: [1814:0801] type 01 class 0x060400
[    0.787908] pci 0000:00:00.0: reg 0x10: [mem 0x00000000-0x7fffffff]
[    0.787923] pci 0000:00:00.0: reg 0x14: [mem 0x00000000-0x0000ffff]
[    0.787999] pci 0000:00:00.0: supports D1
[    0.788011] pci 0000:00:00.0: PME# supported from D0 D1 D3hot
[    0.788301] pci 0000:00:00.0: bridge configuration invalid ([bus 00-00]), reconfiguring
[    0.804314] pci 0000:01:00.0: [168c:003c] type 00 class 0x028000
[    0.804358] pci 0000:01:00.0: reg 0x10: [mem 0x00000000-0x001fffff 64bit]
[    0.804406] pci 0000:01:00.0: reg 0x30: [mem 0x00000000-0x0000ffff pref]
[    0.804504] pci 0000:01:00.0: supports D1
[    0.804516] pci 0000:01:00.0: PME# supported from D0 D1 D3hot
[    0.804740] pci_bus 0000:01: busn_res: [bus 01-ff] end is updated to 01
[    0.804760] pci_bus 0000:00: busn_res: [bus 00-ff] end is updated to 01
[    0.804811] pci 0000:00:00.0: BAR 0: no space for [mem size 0x80000000]
[    0.817860] pci 0000:00:00.0: BAR 0: failed to assign [mem size 0x80000000]
[    0.831681] pci 0000:00:00.0: BAR 8: assigned [mem 0x20000000-0x201fffff]
[    0.845180] pci 0000:00:00.0: BAR 9: assigned [mem 0x20200000-0x202fffff pref]
[    0.859544] pci 0000:00:00.0: BAR 1: assigned [mem 0x20300000-0x2030ffff]
[    0.873052] pci 0000:01:00.0: BAR 0: assigned [mem 0x20000000-0x201fffff 64bit]
[    0.887595] pci 0000:01:00.0: BAR 6: assigned [mem 0x20200000-0x2020ffff pref]
[    0.901946] pci 0000:00:00.0: PCI bridge to [bus 01]
[    0.911802] pci 0000:00:00.0:   bridge window [mem 0x20000000-0x201fffff]
[    0.925315] pci 0000:00:00.0:   bridge window [mem 0x20200000-0x202fffff pref]
[    0.939694] pci 0000:00:00.0: card - bus=0x0, slot = 0x0 irq=0
[    0.951280] pci 0000:01:00.0: card - bus=0x1, slot = 0x0 irq=4
[    0.965999] clocksource: Switched to clocksource systick
[    0.977750] NET: Registered protocol family 2
[    0.987208] TCP established hash table entries: 1024 (order: 0, 4096 bytes)
[    1.000979] TCP bind hash table entries: 1024 (order: 0, 4096 bytes)
[    1.013574] TCP: Hash tables configured (established 1024 bind 1024)
[    1.026298] UDP hash table entries: 256 (order: 0, 4096 bytes)
[    1.037828] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes)
[    1.050515] NET: Registered protocol family 1
[    1.059133] PCI: CLS 80 bytes, default 32
[    1.061357] rt-timer 10000100.timer: maximum frequency is 1220Hz
[    1.074044] Crashlog allocated RAM at address 0x3f00000
[    1.085487] workingset: timestamp_bits=30 max_order=14 bucket_order=0
[    1.106446] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    1.117924] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    1.145828] io scheduler noop registered
[    1.153528] io scheduler deadline registered (default)
[    1.163906] ralink-usb-phy usbphy: invalid resource
[    1.174073] Serial: 8250/16550 driver, 2 ports, IRQ sharing disabled
[    1.187681] console [ttyS0] disabled
[    1.194670] 10000c00.uartlite: ttyS0 at MMIO 0x10000c00 (irq = 20, base_baud = 2500000) is a Palmchip BK-3103
[    1.214440] console [ttyS0] enabled
[    1.228218] bootconsole [early0] disabled
[    1.248404] spi spi0.0: force spi mode3
[    1.256900] m25p80 spi0.0: mx25l6405d (8192 Kbytes)
[    1.266737] 8 ofpart partitions found on MTD device spi0.0
[    1.277666] Creating 8 MTD partitions on "spi0.0":
[    1.287220] 0x000000000000-0x000000030000 : "u-boot"
[    1.298777] 0x000000030000-0x000000040000 : "config"
[    1.310568] 0x000000048000-0x000000050000 : "art"
[    1.321848] 0x000000040000-0x000000050000 : "factory"
[    1.333798] 0x000000050000-0x00000071a000 : "firmware"
[    1.936517] 2 uimage-fw partitions found on MTD device firmware
[    1.948354] 0x000000050000-0x00000019a4c2 : "kernel"
[    1.959823] 0x00000019a4c2-0x00000071a000 : "rootfs"
[    1.971560] mtd: device 6 (rootfs) set to be root filesystem
[    1.983490] 1 squashfs-split partitions found on MTD device rootfs
[    1.995877] 0x0000003cc000-0x00000071a000 : "rootfs_data"
[    2.008214] 0x00000071a000-0x0000007b0000 : "manufacture"
[    2.020865] 0x0000007b0000-0x0000007c0000 : "backup"
[    2.032644] 0x0000007c0000-0x000000800000 : "storage"
[    2.045254] libphy: Fixed MDIO Bus: probed
[    2.057379] libphy: mdio: probed
[    2.074436] mtk_soc_eth 10100000.ethernet: connected port 4 to PHY at mdio-bus:04 [uid=004dd072, driver=Atheros 8035 ethernet]
[    2.097246] mtk_soc_eth 10100000.ethernet: loaded mt7620 driver
[    2.109912] mtk_soc_eth 10100000.ethernet eth0: mediatek frame engine at 0xb0100000, irq 5
[    2.126956] rt2880_wdt 10000120.watchdog: Initialized
[    2.138195] NET: Registered protocol family 10
[    2.150115] NET: Registered protocol family 17
[    2.159096] 8021q: 802.1Q VLAN Support v1.8
[    2.179366] VFS: Mounted root (squashfs filesystem) readonly on device 31:6.
[    2.194421] Freeing unused kernel memory: 168K
[    2.203265] This architecture does not have kernel memory protection.
[    2.804319] random: fast init done
[    3.520946] init: Console is alive
[    3.528041] init: - watchdog -
[    5.302689] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    5.570819] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    5.587490] init: - preinit -
[    7.873205] 8021q: adding VLAN 0 to HW filter on device eth0
[    8.314445] mtk_soc_eth 10100000.ethernet eth0: port 4 link up (1000Mbps/Full duplex)
[    8.352740] mount_root: jffs2 not ready yet, using temporary tmpfs overlay
[    8.394841] urandom-seed: Seed file not found (/etc/urandom.seed)
[    8.441016] procd: - early -
[    8.447631] procd: - watchdog -
[    8.648741] procd: - watchdog -
[    8.655364] procd: - ubus -
[    8.821271] procd: - init -
[    8.904717] mtk_soc_eth 10100000.ethernet eth0: port 4 link down
[    9.258545] kmodloader: loading kernel modules from /etc/modules.d/*
[    9.277450] ip6_tables: (C) 2000-2006 Netfilter Core Team
[    9.301341] Loading modules backported from Linux version wt-2017-11-01-0-gfe248fc2c180
[    9.317363] Backport generated by backports.git v4.14-rc2-1-31-g86cf0e5d
[    9.335367] ip_tables: (C) 2000-2006 Netfilter Core Team
[    9.354815] nf_conntrack version 0.5.0 (1024 buckets, 4096 max)
[    9.424633] xt_time: kernel timezone is -0000
[    9.604921] PPP generic driver version 2.4.2
[    9.616305] NET: Registered protocol family 24
[    9.660269] ath10k_pci 0000:01:00.0: pci irq legacy oper_irq_mode 1 irq_mode 0 reset_mode 0
[    9.790499] ath10k_pci 0000:01:00.0: Direct firmware load for ath10k/pre-cal-pci-0000:01:00.0.bin failed with error -2
[    9.811859] ath10k_pci 0000:01:00.0: Falling back to user helper
[   10.043181] firmware ath10k!pre-cal-pci-0000:01:00.0.bin: firmware_loading_store: map pages failed
[   10.061427] ath10k_pci 0000:01:00.0: Direct firmware load for ath10k/cal-pci-0000:01:00.0.bin failed with error -2
[   10.082097] ath10k_pci 0000:01:00.0: Falling back to user helper
[   10.333472] ath10k_pci 0000:01:00.0: Direct firmware load for ath10k/QCA988X/hw2.0/firmware-6.bin failed with error -2
[   10.354865] ath10k_pci 0000:01:00.0: Falling back to user helper
[   10.463643] firmware ath10k!QCA988X!hw2.0!firmware-6.bin: firmware_loading_store: map pages failed
[   10.715046] ath10k_pci 0000:01:00.0: qca988x hw2.0 target 0x4100016c chip_id 0x043222ff sub 0000:0000
[   10.733511] ath10k_pci 0000:01:00.0: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 1
[   10.756162] ath10k_pci 0000:01:00.0: firmware ver 10.2.4-1.0-00033 api 5 features no-p2p,raw-mode,mfp,allows-mesh-bcast crc32 c41417d0
[   10.813538] ath10k_pci 0000:01:00.0: Direct firmware load for ath10k/QCA988X/hw2.0/board-2.bin failed with error -2
[   10.834363] ath10k_pci 0000:01:00.0: Falling back to user helper
[   10.915014] firmware ath10k!QCA988X!hw2.0!board-2.bin: firmware_loading_store: map pages failed
[   10.932945] ath10k_pci 0000:01:00.0: board_file api 1 bmi_id N/A crc32 bebc7c08
[   12.031173] ath10k_pci 0000:01:00.0: htt-ver 2.1 wmi-op 5 htt-op 2 cal file max-sta 128 raw 0 hwcrypto 1
[   12.103569] ath: EEPROM regdomain: 0x0
[   12.103577] ath: EEPROM indicates default country code should be used
[   12.103580] ath: doing EEPROM country->regdmn map search
[   12.103592] ath: country maps to regdmn code: 0x3a
[   12.103598] ath: Country alpha2 being used: US
[   12.103602] ath: Regpair used: 0x3a
[   12.145975] ieee80211 phy1: rt2x00lib_request_eeprom_file: Info - Loading EEPROM data from 'soc_wmac.eeprom'.
[   12.261492] rt2800_wmac 10180000.wmac: Direct firmware load for soc_wmac.eeprom failed with error -2
[   12.279747] rt2800_wmac 10180000.wmac: Falling back to user helper
[   13.037095] ieee80211 phy1: rt2x00_set_rt: Info - RT chipset 6352, rev 0500 detected
[   13.052624] ieee80211 phy1: rt2x00_set_rf: Info - RF chipset 7620 detected
[   13.067012] ieee80211 phy1: Selected rate control algorithm 'minstrel_ht'
[   13.101179] kmodloader: done loading kernel modules from /etc/modules.d/*
[   19.613355] jffs2_scan_eraseblock(): End of filesystem marker found at 0x0
[   19.652381] jffs2_build_filesystem(): unlocking the mtd device... 
[   19.664374] done.
[   19.668200] jffs2_build_filesystem(): erasing all blocks after the end marker... 
[   23.427207] 8021q: adding VLAN 0 to HW filter on device eth0
[   23.469214] br-lan: port 1(eth0) entered blocking state
[   23.479707] br-lan: port 1(eth0) entered disabled state
[   23.490599] device eth0 entered promiscuous mode
[   23.612274] IPv6: ADDRCONF(NETDEV_UP): br-lan: link is not ready
[   25.875886] br-lan: port 1(eth0) entered blocking state
[   25.886331] br-lan: port 1(eth0) entered forwarding state
[   25.976403] IPv6: ADDRCONF(NETDEV_CHANGE): br-lan: link becomes ready
[   26.583752] mtk_soc_eth 10100000.ethernet eth0: port 4 link up (1000Mbps/Full duplex)
[   55.502215] done.
[   55.506094] jffs2: notice: (1030) jffs2_build_xattr_subsystem: complete building xattr subsystem, 0 of xdatum (0 unchecked, 0 orphan) and 0 of xref (0 dead, 0 orphan) found.
[   88.644843] random: crng init done
`

mkresin commented:
Thanks a lot for your patience!

It has been turned out that the V7 APAC1200 is a rebranded [edimax cap1200](http://www.edimax.com/edimax/merchandise/merchandise_detail/data/edimax/global/edimax_pro_indoor_access_points_ac1200/cap1200/).

If you search the web for the Reg. No. EW-7476HPC on the back of the APAC1200s case, you might find the [Edimax test report on the fcc site](https://fccid.io/NDD9574761413/Test-Report/test-report-2507568).

The [Acelink EW-7476HPC](https://wikidevi.com/wiki/Acelink_EW-7476HPC) is the same board with a different vendor branding. According to the [internal photos](https://fccid.io/NDD9574761413/Internal-Photos/int-photos-2507621) on the fcc site, the board is available as 8MByte and 16MByte version.

Long story short, would you please rename the board to `edimax,cap1200-8m` and the firmware utility to something with edimax in the name. I only had a brief look, but it looks like the header of the edimax firmware is the same. Feel free to mention the V7 APAC1200 in DEVICE_TITLE.

obsy commented:
edimax cap1200 has beeper on & gpio2: gpio12. Please check this on V7 APAC1200

jongerenchaos commented:
Is there a test firmware file that i can install on this device?

mkresin commented:
> Am I missing something?

You need to set the pin 52 to function GPIO. Pin 52 is part of the nd_sd group:

```
&pinctrl {
	state_default: pinctrl0 {
		gpio {
			ralink,group = "i2c", "nd_sd", "uartf";
			ralink,function = "gpio";
		};
	};
};
```

obsy commented:
Exactly.

mkresin commented on target/linux/ramips/dts/APAC1200.dts:
```diff
@@ -0,0 +1,178 @@
+	mtd-mac-address = <&factory 0x4>;
+	mtd-mac-address-increment = <1>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&rgmii1_pins &rgmii2_pins &mdio_pins>;
```
One of the rgmii pins groups here is superfluous, since you're only using one rgmii out of two. But I've no idea which one is used. You would need to set the rgmii[1|2] group to function gpio to see which one is related/required.

eljefe80 replied:
I think I understand what you're asking here, but when I set either of the rgmii groups to function gpio, the switch fails to load:

'''
&pinctrl {
        state_default: pinctrl0 {
                gpio {
                        ralink,group = "i2c", "nd_sd", "rgmii1", "uartf";
                        ralink,function = "gpio";
                };
        };
};
'''
results in:
'''
[    1.840523] rt2880-pinmux pinctrl: pin io24 already requested by pinctrl; cannot claim for 10100000.ethernet
[    1.860171] rt2880-pinmux pinctrl: pin-24 (10100000.ethernet) status -22
[    1.873537] rt2880-pinmux pinctrl: could not request pin 24 (io24) from group rgmii1  on device rt2880-pinmux
[    1.893300] mtk_soc_eth 10100000.ethernet: Error applying setting, reverse things back
[    1.909095] mtk_soc_eth: probe of 10100000.ethernet failed with error -22
'''
and
'''
&pinctrl {
        state_default: pinctrl0 {
                gpio {
                        ralink,group = "i2c", "nd_sd", "rgmii1", "uartf";
                        ralink,function = "gpio";
                };
        };
};
'''
results in
'''
[    1.840542] rt2880-pinmux pinctrl: pin io60 already requested by pinctrl; cannot claim for 10100000.ethernet
[    1.860188] rt2880-pinmux pinctrl: pin-60 (10100000.ethernet) status -22
[    1.873553] rt2880-pinmux pinctrl: could not request pin 60 (io60) from group rgmii2  on device rt2880-pinmux
[    1.893315] mtk_soc_eth 10100000.ethernet: Error applying setting, reverse things back
[    1.909122] mtk_soc_eth: probe of 10100000.ethernet failed with error -22
'''

mkresin replied:
You need to drop the pingroup set to GPIO from `pinctrl-0 = <&rgmii1_pins &rgmii2_pins &mdio_pins>;` as well.

mkresin commented on target/linux/ramips/dts/APAC1200.dts:
```diff
@@ -0,0 +1,178 @@
+			label = "storage";
+			reg = <0x7c0000 0x40000>;
+		};
+
```
drop the empty line here please.

mkresin commented on target/linux/ramips/dts/APAC1200.dts:
```diff
@@ -0,0 +1,178 @@
+			reg = <4>;
+			phy-mode = "rgmii";
+		};
+
```
drop the empty line here please.

mkresin commented on target/linux/ramips/dts/APAC1200.dts:
```diff
@@ -0,0 +1,178 @@
+	mtd-mac-address = <&config 0x29>;
+};
+
+
```
one linebreak is enough

jongerenchaos replied:
Allright i compiled the source and it is running now on my V7 accespoint (with latest kernel 4.9.82)

For now there is one big problem: the 2.4Ghz doesn't work well. The bitrate is very low (1/2Mbit down /up) and  i get some errors:
"ieee80211 phy1: rt2800_config_channel: Warning - Using incomplete support for external PA"

5Ghz no problem, works great!

How can i resolve it?

jongerenchaos commented:
All patches works well with the current git openwrt main version. I compiled the source and it is running now on my V7 accespoint (with latest kernel 4.9.82)

For now there is one big problem: the 2.4Ghz doesn't work well. The bitrate is very low (1/2Mbit down /up) and i get some errors:
"ieee80211 phy1: rt2800_config_channel: Warning - Using incomplete support for external PA"

5Ghz no problem, works great!

How can i resolve it?

obsy commented on tools/firmware-utils/src/mkv7apac.c:
```diff
@@ -0,0 +1,188 @@
+			case 'o':
+				ofn = optarg;
+				break;
+			case 'd':
```
Missing "-d"  option. Typo in line 114

mkresin commented:
I added the PR to [my staging tree](https://git.openwrt.org/?p=openwrt/staging/mkresin.git;a=summary) but it needs extensive testing, since I changed a lot of stuff.

1) board dropped from 10-rt2x00-eeprom and switched to ralink,mtd-eeprom device tree property
2) unused functions from 11-ath10k-caldata dropped
3) alphabetical order issues are fixed
4) mkv7apac.c dropped
5) all issues found by checkpatch for mkedimax-cap1200.c fixed
6) replaced mentions of mkv7apac with edimax-cap1200
7) decode option removed from mkedimax-cap1200.c as there is no code to decode the image
8) compiler warnings enabled for mkedimax-cap1200.c enabled
9) most compiler warning for mkedimax-cap1200.c fixed

Here is where I finally ran out of time. There are still the following compiler warning, which I would like to see fixed:

```
src/mkedimax-cap1200.c: In function 'main':
src/mkedimax-cap1200.c:146:32: warning: pointer targets in passing argument 2 of 'write_header' differ in signedness [-Wpointer-sign]
  if (ofn && (write_header(out, md5sum))) {
                                ^
src/mkedimax-cap1200.c:56:5: note: expected 'char *' but argument is of type 'uint8_t * {aka unsigned char *}'
 int write_header(FILE *out, char *md5sum) {
     ^
```

I wouldn't be surprised if the wireless mac address doesn't match the one used by the stock firmware. It can be fixed with the devicetree properties `mtd-mac-address` + `mtd-mac-address-increment`. For ath10k it can be patched into the caldata.

What ever you do, please update the PR with what I have in [my staging tree](https://git.openwrt.org/?p=openwrt/staging/mkresin.git;a=summary). I will not invest that much time again.

mkresin commented:
I unstaged your commit from my staging tree. If you have the time to fix the remaining issues, please use the following commit as base and reopen the PR.

```
From c706fb415ce8f3f5eeefc03d1e39f0b6d1fc8aca Mon Sep 17 00:00:00 2001
From: Jeff Williams <jeff@wdwconsulting.net>
Date: Mon, 29 Jan 2018 16:16:10 -0800
Subject: [PATCH] ramips: add support for Edimax CAP1200

The patch adds support for the Edimax CAP1200, also known as
V7 APAC1200 and Acelink EW-7476HPC.

Specifications:
  * SoC: 580Mhz MT7620a
  * RAM: 64MB
  * Storage: 8MB of FLASH (SPI NOR)
  * 1x ar8035 GB switch
  * Ethernet: 1x1000M (802.11at POE), port #4 on ar8035
  * Wireless: QCA988X 802.11ac
              MT7620 2.4GHz 802.11b/g/n
  * 2x GPIO LED
  * 1x GPIO Reset Button
  * 1x DC Jack 12v
  * 1x UART, 3.3v, 57600

Installation:

   There are 2 ways of installation:

Serial Port/TFTP

1. Setup a tftp server and copy/link an initramfs image

 $ sudo ifconfig eth0 192.168.99.9 netmask 255.255.255.0
 $ cp openwrt-ramips-mt7620-edimax_cap1200-8m-initramfs-kernel.bin \
   /srv/tftp/uImage_EW-7476HPC

2. Connect via ttyl uart:

 On the board is a 4-pin, 3.3v uart running at 115200bps, 8n1, etc.

3. For a few seconds after power on and during reboot, you can hit 1 to
   break the automatic boot sequence and it will prompt for tftp settings,
   use the default. This will boot the openwrt initramfs kernel.

4. Change your ip address to 192.168.1.2

 $ sudo ifconfig eth0 192.168.1.2 netmask 255.255.255.0

5. Upload the sysupgrade image to the board with scp:

 $ scp openwrt-ramips-mt7620-edimax_cap1200-8m-squashfs-sysupgrade.bin \
 root@192.168.1.1:/tmp/fw.bin

6. Log in to the running system listening on 192.168.1.1 via ssh
   as root (without password):

 $ ssh root@192.168.1.1

7. Flash the uploaded firmware file from the ssh session via the
   sysupgrade command:

 root@OpenWrt:~# sysupgrade /tmp/fw.bin

Web interface

1. Navigate to the upgrade page on the device and upload the file

  openwrt-ramips-mt7620-edimax_cap1200-8m-squashfs-factory.bin

Signed-off-by: Jeff Williams <jeff@wdwconsulting.net>
---
 target/linux/ramips/base-files/etc/board.d/01_leds |   4 +
 .../linux/ramips/base-files/etc/board.d/02_network |   1 +
 .../etc/hotplug.d/firmware/10-rt2x00-eeprom        |   3 +
 .../etc/hotplug.d/firmware/11-ath10k-caldata       |  41 +++++
 .../ramips/base-files/lib/upgrade/platform.sh      |   1 +
 target/linux/ramips/dts/CAP1200.dts                | 180 ++++++++++++++++++++
 target/linux/ramips/image/mt7620.mk                |  16 ++
 tools/firmware-utils/Makefile                      |   1 +
 tools/firmware-utils/src/mkedimax-cap1200.c        | 182 +++++++++++++++++++++
 9 files changed, 429 insertions(+)
 create mode 100644 target/linux/ramips/base-files/etc/hotplug.d/firmware/11-ath10k-caldata
 create mode 100644 target/linux/ramips/dts/CAP1200.dts
 create mode 100644 tools/firmware-utils/src/mkedimax-cap1200.c

diff --git a/target/linux/ramips/base-files/etc/board.d/01_leds b/target/linux/ramips/base-files/etc/board.d/01_leds
index dfc56b0..df022d6 100755
--- a/target/linux/ramips/base-files/etc/board.d/01_leds
+++ b/target/linux/ramips/base-files/etc/board.d/01_leds
@@ -176,6 +176,10 @@ vr500)
 dir-860l-b1)
 	ucidef_set_led_netdev "wan" "wan" "$boardname:green:net" "eth0.2"
 	;;
+edimax,cap1200-8m)
+	set_wifi_led "$boardname:yellow:wlan5g" "wlan0"
+	set_wifi_led "$boardname:blue:wlan2g" "wlan1"
+	;;
 ex2700|\
 wn3000rpv3)
 	set_wifi_led "$boardname:green:router"
diff --git a/target/linux/ramips/base-files/etc/board.d/02_network b/target/linux/ramips/base-files/etc/board.d/02_network
index b7e5fb6..35a38b6 100755
--- a/target/linux/ramips/base-files/etc/board.d/02_network
+++ b/target/linux/ramips/base-files/etc/board.d/02_network
@@ -134,6 +134,7 @@ ramips_setup_interfaces()
 	d105|\
 	dlink,dap-1522-a1|\
 	dch-m225|\
+	edimax,cap1200-8m|\
 	ex2700|\
 	ex3700|\
 	hpm|\
diff --git a/target/linux/ramips/base-files/etc/hotplug.d/firmware/10-rt2x00-eeprom b/target/linux/ramips/base-files/etc/hotplug.d/firmware/10-rt2x00-eeprom
index c0c6a3e..447a040 100644
--- a/target/linux/ramips/base-files/etc/hotplug.d/firmware/10-rt2x00-eeprom
+++ b/target/linux/ramips/base-files/etc/hotplug.d/firmware/10-rt2x00-eeprom
@@ -61,6 +61,9 @@ case "$FIRMWARE" in
 		jboot_eeprom_extract "config" 0xE000
 		rt2x00_eeprom_set_macaddr $wifi_mac
 		;;
+	edimax,cap1200-8m)
+		rt2x00_eeprom_extract "factory" 0 512
+		;;
 	tiny-ac)
 		wifi_mac=$(mtd_get_mac_ascii u-boot-env INIC_MAC_ADDR)
 		rt2x00_eeprom_extract "factory" 0 512
diff --git a/target/linux/ramips/base-files/etc/hotplug.d/firmware/11-ath10k-caldata b/target/linux/ramips/base-files/etc/hotplug.d/firmware/11-ath10k-caldata
new file mode 100644
index 0000000..0123adc
--- /dev/null
+++ b/target/linux/ramips/base-files/etc/hotplug.d/firmware/11-ath10k-caldata
@@ -0,0 +1,41 @@
+#!/bin/sh
+
+ath10kcal_die() {
+	echo "ath10cal: " "$*"
+	exit 1
+}
+
+ath10kcal_extract() {
+	local part=$1
+	local offset=$2
+	local count=$3
+	local mtd
+
+	mtd=$(find_mtd_chardev $part)
+	[ -n "$mtd" ] || \
+		ath10kcal_die "no mtd device found for partition $part"
+
+	dd if=$mtd of=/lib/firmware/$FIRMWARE bs=1 skip=$offset count=$count 2>/dev/null || \
+		ath10kcal_die "failed to extract calibration data from $mtd"
+}
+
+[ -e /lib/firmware/$FIRMWARE ] && exit 0
+
+. /lib/functions.sh
+. /lib/functions/system.sh
+
+board=$(board_name)
+
+
+case "$FIRMWARE" in
+"ath10k/cal-pci-0000:01:00.0.bin")
+	case "$board" in
+	edimax,cap1200-8m)
+		ath10kcal_extract "art" 0 2116
+		;;
+	esac
+	;;
+*)
+	exit 1
+	;;
+esac
diff --git a/target/linux/ramips/base-files/lib/upgrade/platform.sh b/target/linux/ramips/base-files/lib/upgrade/platform.sh
index 0a43f29..52d60b7 100755
--- a/target/linux/ramips/base-files/lib/upgrade/platform.sh
+++ b/target/linux/ramips/base-files/lib/upgrade/platform.sh
@@ -53,6 +53,7 @@ platform_check_image() {
 	dir-810l|\
 	duzun-dm06|\
 	e1700|\
+	edimax,cap1200-8m|\
 	esr-9753|\
 	ew1200|\
 	ex2700|\
diff --git a/target/linux/ramips/dts/CAP1200.dts b/target/linux/ramips/dts/CAP1200.dts
new file mode 100644
index 0000000..fa4593d
--- /dev/null
+++ b/target/linux/ramips/dts/CAP1200.dts
@@ -0,0 +1,180 @@
+/*
+ *  BSD LICENSE
+ *
+ *  Copyright (C) 2018 Jeff Williams <jeff@wdwconsulting.net>
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *    1. Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *    2. Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *    3. Neither the names of the copyright holders nor the names of any
+ *       contributors may be used to endorse or promote products derived
+ *       from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ *  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ *  HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ *  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ *  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ *  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/dts-v1/;
+
+#include "mt7620a.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "edimax,cap1200-8m", "ralink,mt7620a-soc";
+	model = "Edimax CAP1200";
+
+	chosen {
+		bootargs = "console=ttyS0,57600";
+	};
+
+	gpio-keys {
+		compatible = "gpio-keys";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		poll-interval = <20>;
+
+		reset {
+			label = "reset";
+			gpios = <&gpio0 2 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+	};
+
+	gpio-leds {
+		compatible = "gpio-leds";
+
+		wlan2g {
+			label = "cap1200:blue:wlan2g";
+			gpios = <&gpio0 1 GPIO_ACTIVE_LOW>;
+		};
+
+		wlan5g {
+			label = "cap1200:yellow:wlan5g";
+			gpios = <&gpio0 12 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	beeper: beeper {
+		compatible = "gpio-beeper";
+		gpios = <&gpio2 12 GPIO_ACTIVE_LOW>;
+	};
+};
+
+&spi0 {
+	status = "okay";
+
+	m25p80@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
+
+		partition@0 {
+			label = "u-boot";
+			reg = <0x0 0x30000>;
+			read-only;
+		};
+
+		config: partition@30000 {
+			label = "config";
+			reg = <0x30000 0x10000>;
+			read-only;
+		};
+
+		art: partition@48000 {
+			label = "art";
+			reg = <0x48000 0x8000>;
+			read-only;
+		};
+
+		factory: partition@40000 {
+			label = "factory";
+			reg = <0x40000 0x10000>;
+			read-only;
+		};
+
+		partition@50000 {
+			label = "firmware";
+			reg = <0x50000 0x6ca000>;
+		};
+
+		partition@71a000 {
+			label = "manufacture";
+			reg = <0x71a000 0x96000>;
+		};
+
+		partition@7b0000 {
+			label = "backup";
+			reg = <0x7b0000 0x10000>;
+		};
+
+		partition@7c0000 {
+			label = "storage";
+			reg = <0x7c0000 0x40000>;
+		};
+	};
+};
+
+&pinctrl {
+	state_default: pinctrl0 {
+		gpio {
+			ralink,group = "i2c", "nd_sd", "uartf";
+			ralink,function = "gpio";
+		};
+	};
+};
+
+&ethernet {
+	status = "okay";
+	mtd-mac-address = <&factory 0x4>;
+	mtd-mac-address-increment = <1>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&rgmii1_pins &mdio_pins>;
+
+	port@4 {
+		status = "okay";
+		phy-mode = "rgmii";
+		phy-handle = <&phy4>;
+	};
+
+	mdio-bus {
+		status = "okay";
+
+		phy4: ethernet-phy@4 {
+			reg = <4>;
+			phy-mode = "rgmii";
+		};
+	};
+};
+
+&gsw {
+	mediatek,port4 = "gmac";
+	mtd-mac-address = <&config 0x29>;
+};
+
+&pcie {
+	status = "okay";
+};
+
+&wmac {
+	ralink,mtd-eeprom = <&factory 0>;
+};
diff --git a/target/linux/ramips/image/mt7620.mk b/target/linux/ramips/image/mt7620.mk
index c12a915..ddc3793 100644
--- a/target/linux/ramips/image/mt7620.mk
+++ b/target/linux/ramips/image/mt7620.mk
@@ -5,6 +5,11 @@
 DEVICE_VARS += TPLINK_FLASHLAYOUT TPLINK_HWID TPLINK_HWREV TPLINK_HWREVADD TPLINK_HVERSION \
 	DLINK_ROM_ID DLINK_FAMILY_MEMBER DLINK_FIRMWARE_SIZE
 
+define Build/edimax_cap1200
+	$(STAGING_DIR_HOST)/bin/mkedimax-cap1200 -i $@ -o $@.xor
+	mv -f $@.xor $@
+endef
+
 define Build/elecom-header
 	cp $@ $(KDIR)/v_0.0.0.bin
 	( \
@@ -194,6 +199,17 @@ define Device/e1700
 endef
 TARGET_DEVICES += e1700
 
+define Device/edimax_cap1200-8m
+  DTS := CAP1200
+  DEVICE_TITLE := Edimax CAP1200
+  DEVICE_PACKAGES := kmod-ath10k ath10k-firmware-qca988x
+  IMAGE_SIZE := 7118848
+  IMAGES += factory.bin
+  IMAGE/factory.bin := $$(sysupgrade_bin) | pad-to 64k | \
+	check-size $$$$(IMAGE_SIZE) | edimax_cap1200
+endef
+TARGET_DEVICES += edimax_cap1200-8m
+
 define Device/ex2700
   NETGEAR_HW_ID := 29764623+4+0+32+2x2+0
   NETGEAR_BOARD_ID := EX2700
diff --git a/tools/firmware-utils/Makefile b/tools/firmware-utils/Makefile
index 7c0180a..71b99dc 100644
--- a/tools/firmware-utils/Makefile
+++ b/tools/firmware-utils/Makefile
@@ -24,6 +24,7 @@ define Host/Compile
 	$(call cc,motorola-bin)
 	$(call cc,dgfirmware)
 	$(call cc,mksenaofw md5)
+	$(call cc,mkedimax-cap1200 md5, -Wall)
 	$(call cc,trx2usr)
 	$(call cc,ptgen)
 	$(call cc,srec2bin)
diff --git a/tools/firmware-utils/src/mkedimax-cap1200.c b/tools/firmware-utils/src/mkedimax-cap1200.c
new file mode 100644
index 0000000..8651d7a
--- /dev/null
+++ b/tools/firmware-utils/src/mkedimax-cap1200.c
@@ -0,0 +1,182 @@
+/*
+ * mkedimax-cap1200.c - partially based on OpenWrt's xorimage.c
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ */
+
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdint.h>
+#include <unistd.h>
+#include <sys/stat.h>
+#include "md5.h"
+
+#define BUF_SIZE	0x200
+
+static uint _pattern = 0x30471688;
+
+static int md5_file(const char *filename, uint8_t *dst)
+{
+	FILE *fp_src;
+	MD5_CTX ctx;
+	char buf[BUF_SIZE];
+	size_t bytes_read;
+
+	MD5_Init(&ctx);
+
+	fp_src = fopen(filename, "r+b");
+	if (!fp_src)
+		return -1;
+
+	while (!feof(fp_src)) {
+		bytes_read = fread(&buf, 1, BUF_SIZE, fp_src);
+		MD5_Update(&ctx, &buf, bytes_read);
+	}
+	fclose(fp_src);
+
+	MD5_Final(dst, &ctx);
+
+	return 0;
+}
+
+int write_header(FILE *out, char *md5sum) {
+	char header1[70] = {
+		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x03,
+		0x47, 0x45, 0x54, 0x5f, 0x53, 0x54, 0x41, 0x47,
+		0x49, 0x4e, 0x47, 0x2f, 0x61, 0x70, 0x70, 0x73,
+		0x5f, 0x31, 0x2e, 0x30, 0x2e, 0x31, 0x30, 0x00,
+		0x00, 0x00, 0x00, 0x00,	0x00, 0x00, 0x00, 0x00,
+		0x00, 0x00, 0xff, 0xff, 0x00, 0x12, 0x33, 0x30,
+		0x34, 0x37, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x65,
+		0xf0, 0x00, 0x00, 0x00, 0x00, 0x00 };
+	char header2[42] = {
+		0xdd, 0xbf, 0x94, 0x1a, 0x02, 0x40, 0x00, 0x00,
+		0x00, 0x00, 0x48, 0x4b, 0x02, 0x40, 0x01, 0x00,
+		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00,
+		0x00, 0x00, 0x38, 0x19, 0x02, 0x40, 0x00, 0x00,
+		0x00, 0x00, 0x00, 0x00,	0x07, 0x18, 0x30, 0x47,
+		0x16, 0x88 };
+
+	if (!fwrite(header1, sizeof(header1), 1, out)) {
+		fprintf(stderr, "fwrite error\n");
+		return EXIT_FAILURE;
+	}
+
+	if (!fwrite(md5sum, 0x10, 1, out)) {
+		fprintf(stderr, "fwrite error\n");
+		return EXIT_FAILURE;
+	}
+
+	if (!fwrite(header2, sizeof(header2), 1, out)) {
+		fprintf(stderr, "fwrite error\n");
+		return EXIT_FAILURE;
+	}
+	return 0;
+}
+
+void usage(void) __attribute__ ((__noreturn__));
+
+void usage(void)
+{
+	fprintf(stderr,
+		"Usage: mkedimax-cap1200 [-i infile] [-o outfile] [-d]\n");
+	exit(EXIT_FAILURE);
+}
+
+int main(int argc, char **argv)
+{
+	char buf[1];	/* keep this at 1k or adjust garbage calc below */
+	FILE *in = stdin;
+	FILE *out = stdout;
+	char *ifn = NULL;
+	char *ofn = NULL;
+	int c, count;
+	size_t n;
+	uint8_t md5sum[0x10];
+
+	while ((c = getopt(argc, argv, "i:o:h")) != -1) {
+		switch (c) {
+		case 'i':
+			ifn = optarg;
+			break;
+		case 'o':
+			ofn = optarg;
+			break;
+		case 'h':
+		default:
+			usage();
+		}
+	}
+
+	if (optind != argc || optind == 1) {
+		fprintf(stderr, "illegal arg \"%s\"\n", argv[optind]);
+		usage();
+	}
+
+	if (ifn && md5_file(ifn, md5sum)) {
+		fprintf(stderr, "can not get md5sum for \"%s\"\n", ifn);
+		usage();
+	}
+
+	if (ifn && !(in = fopen(ifn, "r"))) {
+		fprintf(stderr, "can not open \"%s\" for reading\n", ifn);
+		usage();
+	}
+
+	if (ofn && !(out = fopen(ofn, "w"))) {
+		fprintf(stderr, "can not open \"%s\" for writing\n", ofn);
+		usage();
+	}
+
+	if (ofn && (write_header(out, md5sum))) {
+		fprintf(stderr, "can not write header to \"%s\"\n", ofn);
+		usage();
+	}
+
+
+	count = 0;
+	while ((n = fread(buf, 1, sizeof(buf), in)) > 0) {
+		if (n < sizeof(buf) && ferror(in)) {
+			fprintf(stderr, "fread error\n");
+			return EXIT_FAILURE;
+		}
+
+		buf[0] = (buf[0] ^ (_pattern >> count)) & 0xff;
+
+		if (!fwrite(buf, n, 1, out)) {
+			fprintf(stderr, "fwrite error\n");
+			return EXIT_FAILURE;
+		}
+		count = (count + 1) % 8;
+	}
+
+	if (ferror(in)) {
+		fprintf(stderr, "fread error\n");
+		return EXIT_FAILURE;
+	}
+
+	if (fflush(out)) {
+		fprintf(stderr, "fwrite error\n");
+		return EXIT_FAILURE;
+	}
+
+	fclose(in);
+	fclose(out);
+
+	return EXIT_SUCCESS;
+}
-- 
2.7.4

```

locngy commented:
Why this changes haven't merged to openwrt:master?

dmascord commented:
Adapted this in another pull request to support the 16M version - https://github.com/openwrt/openwrt/pull/4234
